### ISSUE TITLE
Implement Clone, Sync and Send

### ISSUE THREAD

**Issue by hghwng** [endoli/lldb.rs#10]
LLDB's stable APIs (`SBxxx`) are thread-safe (see [include/lldb/Target/Target.h](https://github.com/llvm-mirror/lldb/blob/cae2d34a12503d1d3c23172f00cf20bef63641fd/include/lldb/Target/Target.h#L1269)). Opaque objects exported to the outside are also cloneable (take `SBProcess` as an example, see [include/lldb/lldb-forward.h](https://github.com/llvm-mirror/lldb/blob/a063373af4d295e48017ca218d1030aac13b0af5/include/lldb/lldb-forward.h#L408) and [source/API/SBProcess.cpp](https://github.com/llvm-mirror/lldb/blob/d3f69c4e1e016673145715a68a07fc39cf170f94/source/API/SBProcess.cpp#L60)).

It would be nice if `Clone`, `Sync` and `Send` are implemented on these objects.

**Comment by waywardmonkeys** [comment]
Should we do this for everything? Or some subset?

**Comment by hghwng** [comment]
# Clone
I think we should add C bindings for copy constructors in lldb-sys, then implement `Clone` traits manually. For example, the C++ part of `SBDebugger` looks like:
```c++
// In lldb-sys.rs:
SBDebuggerRef SBDebuggerClone(SBDebuggerRef rhs);

// ... which corresponds to the following constructor in lldb/source/API/SBDebugger.cpp
SBDebugger::SBDebugger(const SBDebugger &rhs) : m_opaque_sp(rhs.m_opaque_sp) {}
```
The Rust part:
```rust

impl Clone for SBDebugger {
    fn clone(&self) -> Self {
        SBDebugger {
            raw: unsafe { sys::SBDebuggerClone(self.raw) },
        }
    }
}
```

# Send
I think it's okay to directly implement this trait.

First, the management of internal objects is safe. Exposed objects contain smart pointers to internal objects, and smart pointers in C++ are thread-safe. It's okay to clone them and send them across threads.

Second, the execution of debugger logic is safe — according to the comment I referred above, all `SBxxx` APIs are thread-safe.

# Sync
I think more discussion is needed. Is it safe to access an LLDB object concurrently?

**Comment by waywardmonkeys** [comment]
Thanks for looking into this!

This sounds good / correct to me. I think we can impl `Send` on the `lldb-sys` types and then the types in `lldb` would pick up `Send` automatically, yes?

Do you plan to do some of this work or should I take a look?

**Comment by hghwng** [comment]
Exactly, we can implement `Send` in `lldb-sys`. If things go smoothly, I can start coding by this weekend.

BTW, what about `Sync`?

**Comment by waywardmonkeys** [comment]
I will have to take a look at `Sync` still ... I suspect we should impl `Sync` as well though.

I'm going to be away in the countryside for the next few days. I'll still be checking email and doing some things, but my responses might be slower.

**Comment by hghwng** [comment]
No problem! I'll start working on it soon. Looking forward to your investigation on implementing `Sync`.

**Comment by waywardmonkeys** [comment]
I've pushed changes to `lldb` for `Send` and `Sync`.

**Comment by waywardmonkeys** [comment]
This has been resolved in 339bb5df670c91a3f1545ce96d17cdb62d4c3a83 and 202f0defe27dc61abd6c04820be53eb9dad64205
